wd_util.c

```diff
@@ -2581,3 +2583,63 @@ void wd_alg_attrs_uninit(struct wd_init_attrs *attrs)
 	wd_sched_rr_release(alg_sched);
 }
 
+static int wd_set_ctx_nums(struct wd_ctx_nums *ctx_nums, const char *section,
+			   __u32 op_type_num, int is_comp)
+{
+	int i, j, ctx_num, node, ret;
+	char *ctx_section;
+	const char *type;
+
+	if (is_comp && op_type_num > ARRAY_SIZE(comp_ctx_type))
+		return -WD_EINVAL;
+
+	ctx_section = index(section, ':');
+	ctx_section++;
+	ret = parse_num_on_numa(ctx_section, &ctx_num, &node);
+	if (ret)
+		return ret;
+
+	for (i = 0; i < CTX_MODE_MAX; i++) {
+		for (j = 0; j < op_type_num; j++) {
+			type = is_comp ? comp_ctx_type[i][j] : ctx_type[i][0];
+			if (strncmp(section, type, strlen(type)))
+				continue;
+
+			if (!i)
+				ctx_nums[j].sync_ctx_num = ctx_num;
+			else
+				ctx_nums[j].async_ctx_num = ctx_num;
+
+			return 0;
+		}
+	}
+
+	return -WD_EINVAL;
+}
+
+int wd_env_set_ctx_nums(const char *name, struct wd_ctx_nums *ctx_nums, __u32 size)
+{
+	char *left, *section, *start, *var_s;
+	int ret, is_comp;
+
+	is_comp = strncmp(name, "WD_COMP_CTX_NUM", strlen(name)) ? 0 : 1;
+
```

Comment:
is it because init2 interface requires env to setup ctx, so no need ctx as para?
but wd_xxx_init interface requires ctx as para.

Reply:
WD_COMP_CTX_NUM is not set like others,  it need use comp_ctx_type ("sync-comp:", "sync-decomp:") to match, others use ctx_type({"sync:"}, {"async:"}).
It was meant to parse env first in init2. If env information is right. it will override user's input ctx setup.

Reply:
> WD_COMP_CTX_NUM is not set like others, it need use comp_ctx_type ("sync-comp:", "sync-decomp:") to match, others use ctx_type({"sync:"}, {"async:"}).

Thanks, 
is it possible to change "sync-comp:", "sync-decomp:" to "sync:""async:", then it may be simpler.

> It was meant to parse env first in init2. If env information is right. it will override user's input ctx setup.

How about adding comments to the code. "override user's input ctx setup".

Reply:
> > WD_COMP_CTX_NUM is not set like others, it need use comp_ctx_type ("sync-comp:", "sync-decomp:") to match, others use ctx_type({"sync:"}, {"async:"}).
> 
> Thanks, is it possible to change "sync-comp:", "sync-decomp:" to "sync:""async:", then it may be simpler.
> 
> > It was meant to parse env first in init2. If env information is right. it will override user's input ctx setup.
> 
> How about adding comments to the code. "override user's input ctx setup".

Put wd_env_set_ctx_num inside of wd_ctx_param_init, it's more relevant, and add some comments now.

Thanks.